File: lab4-parser-generator/src/main/java/booleanExpressions/Tree.java
package booleanExpressions;

import guru.nidi.graphviz.attribute.Label;
import guru.nidi.graphviz.engine.Format;
import guru.nidi.graphviz.engine.Graphviz;
import guru.nidi.graphviz.model.MutableGraph;
import guru.nidi.graphviz.model.MutableNode;

import java.io.File;
import java.io.IOException;
import java.util.List;

import static guru.nidi.graphviz.model.Factory.*;

public class Tree {
    private final String value;
    private final LexicalAnalyzer.Token token;
    private final List<Tree> children;
    
    public Tree(String value, List<Tree> treeList) {
        this.value = value;
        this.children = treeList;
        this.token = null;
    }

    public Tree(String value) {
        this.value = value;
        this.children = List.of();
        this.token = null;
    }

    public Tree(LexicalAnalyzer.Token token) {
            if (token.equals(LexicalAnalyzer.EOF)) {
                this.value = "EOF";
            } else {
                this.value = token.body;
            }
            this.token = token;
            this.children = List.of();
        }

    private int getGraphviz(MutableGraph mutableGraph, int nowCount) {
        MutableNode node = mutNode(String.valueOf(nowCount)).add(Label.lines(value));
        var count = nowCount;

        for (Tree child : children) {
            count++;
            mutableGraph.add(
                    node.addLink(mutNode(String.valueOf(count)).add(Label.lines(child.value)))
            );
            count = child.getGraphviz(mutableGraph, count);
        }

        return count;
    }

    private void getGraphviz(MutableGraph mutableGraph) {
        getGraphviz(mutableGraph, 0);
    }

    public void writeGraphviz(String src) throws IOException {
         MutableGraph g = mutGraph("tmp").setDirected(true);

        getGraphviz(g);

        Graphviz.fromGraph(g).render(Format.PNG).toFile(new File(src));
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder;
        if (children.isEmpty() && value != null && !value.equals("EPS")) {
            stringBuilder = new StringBuilder(value);
        } else {
            stringBuilder = new StringBuilder();
        }

        for (Tree child : children) {
            stringBuilder.append(child.toString());
        }

        return stringBuilder.toString();
    }
}

File: lab2-boolean-expression/src/main/kotlin/Tree.kt
import guru.nidi.graphviz.attribute.Label
import guru.nidi.graphviz.engine.Format
import guru.nidi.graphviz.engine.Graphviz
import guru.nidi.graphviz.model.Factory.*
import guru.nidi.graphviz.model.MutableGraph
import java.io.File

data class Tree(private val value: String, private val children: List<Tree> = listOf()) {

    private fun getGraphviz(mutableGraph: MutableGraph, nowCount: Int = 0): Int {
        val node = mutNode(nowCount.toString()).add(Label.lines(value))
        var count = nowCount

        for (child in children) {
            count++
            mutableGraph.add(
                node.addLink(mutNode((count).toString()).add(Label.lines(child.value)))
            )
            count = child.getGraphviz(mutableGraph, count)
        }

        return count
    }

    fun writeGraphviz(src: String) {
        val g = mutGraph("tmp").setDirected(true)

        getGraphviz(g)

        Graphviz.fromGraph(g).render(Format.PNG).toFile(File(src))
    }

    override fun toString(): String {

        val stringBuilder: StringBuilder = if (value !in listOf("S", "E", "T", "P", "S'", "E'", "T'", "eps", "END", "M", "H", "H'")) {
            StringBuilder(value)
        } else {
            StringBuilder()
        }

        for (child in children) {
            stringBuilder.append(child.toString())
        }

        return stringBuilder.toString()
    }
}
